proposals/0000-replacing-cabal-custom-build.md

```diff
@@ -0,0 +1,1378 @@
+  { preBuildComponentHook  :: Maybe BuildComponentHook
+  , postBuildComponentHook :: Maybe BuildComponentHook
+  }
+```
+
+This design ensures that the build hooks are consistently run in all
+build-like phases. This reflects the common pattern in custom Setup scripts
+that one would update the `haddock` and `repl` hooks to mirror the `build` hooks
+(which one can easily forget to do, and end up with an unusable `repl`, for example, see `singletons-base`
+which fails to update the `replHook`).
+
+The build tool executes the build phase as follows:
+
+- Run the `preBuildComponentHook`, giving it access to the specific information
+  about the component being built.  This cannot modify any configuration but may
+  have side effects.
+
+- Build the component.
+
+- Run the `postBuildComponentHook`.
+
+A typical use case for the `preBuildComponentHook` is generating source code for
+modules.
+
+The `postBuildComponentHook` can be used to do further work after building the
+component. For example, in Agda, once the `agda` executable is built a
+post-build hook could run it to compile the builtin libraries and generate
+`.agdai` interface files for them.
+
+There are deliberately no package-level pre-build or post-build hooks, only
+component-level hooks. This avoids introducing unnecessary synchronisation
+points when multiple packages/components are being built in parallel.
+
+
+### Copy hooks
+
+The `copy` hooks run before and after the copy phase, which moves build artifacts
+from the build directory into the install directory.
+
+```haskell
+data CopyComponentInputs
+  = CopyComponentInputs
+  { localBuildInfo :: LocalBuildInfo
+  , copyFlags      :: CopyFlags
+  , targetInfo     :: TargetInfo
+  }
+
+type CopyComponentHook = CopyComponentInputs -> IO ()
+
+data CopyHooks
+  = CopyHooks
+  { preCopyComponentHook  :: Maybe CopyComponentHook
+  , postCopyComponentHook :: Maybe CopyComponentHook
+  }
+```
+
+The copy hooks can be used to copy files per-component.
+The main use case for copy hooks is when set of things you want to install is
+not fixed and predetermined. For example, in the Agda
+example, if `postBuildComponentHook` has generated `.agdai` interface files then
+`preCopyComponentHook` can be used to copy over the generated files when
+installing the compiler.
+
+We could imagine a more declarative way of specifying this being introduced in
+the future, in which case packages will be free to migrate to it gradually.
+There is not necessarily a problem with having some overlap between hooks and
+declarative features.
+
+An alternative approach would be to regard as illegitimate any use cases which
+treat `Cabal` as a packaging and distribution mechanism for executables, and on
+that basis, cease to provide copy hooks.  We do not follow this approach because
+it would significantly inconvenience maintainers of packages that rely on this
+behaviour (e.g. Agda and Darcs), for a relatively small reduction in complexity
+in `Cabal`.
```

Comment:
I'm not pleased that this is dismissed so quickly.

1. I do not buy that the inconvenience is significant without actually talking to then. Changing build instructions from `cabal install agda` to `cabal install agda; ./build-agda-stdlib` is not a big deal at all. **Because no one has `Agda` as a `build-depends` there is no good reason for the cabal build to be self contained.**

2. The point isn't "making Cabal simpler", but opening the door to architectures. What exactly is the semantics of "copy"? Do you need to "copy" the upstream library before the downstream project is built? What should HLS, should doesn't want to install random things during development / incremental builds do for these?

The entire copy phase is a blatant post-hook to building --- the same objections @michaelpj raised to post hooks apply to this. It has no semantics. It cannot be thought of in a demand-driven way. There is no spec for which other build systems like a new HLS one could implement this.